### ISSUE TITLE
Report on Running QA Stuff

### ISSUE THREAD

**Issue by citibeth** [LLNL/spack#1722]
@adamjstewart The new separation of unit, doc and flake8 tests is really nice.  Here's my report on trying to run them, I think another round of improvements is probably needed.

**flake8**
------------
Worked like a charm:
```
spack install py-flake8
spack load py-flake8
cd spack/lib/spack/docs
flake8 my-file.py
```
BTW: I would consider just naming the package flake8, not py-flake8.  Because it's (sort of) a top-level app, the fact that it's implemented in Python is not so important.  Admittedly, this is a grey area.

flake8 reports line numbers but no context.  That makes it really hard to follow and correct its errors off of Travis.  You really need to install flake8 yourself.  Luckily, installing it myself worked.

**Unit tests**
---------------
I tried running unit tests as I always have before (go to the directory, run it with nosetest or python), and that failed miserably.  I see now, there's a qa script.

But that script is not useful for my needs.  To be useful in rapid debugging, we need a way to run a SINGLE unit test, even a SINGLE test case within a single unit test.  We also need a way to run a SINGLE directory of unit tests, plus all unit tests.  Nosetest gives this  out of the box, but that functionality has now been obscured by a single run-all-the-tests script.

Ultimately, I copied out the one unit test I needed to a different directory, put an 'if __main__' at the end, hacked my PYTHONPATH to point to Spack, and ran it with python.  Hack, hack, hack.  This needs to be set up better so I can run single (or a selection of) tests without those kinds of hacks.

**sphinx**
-----------
I could not get Sphinx working.  See console log:
```
vpn-128-154-224-57:docs me$ spack find py-sphinx
==> 1 installed packages.
-- darwin-elcapitan-x86_64 / clang@7.3.0-apple ------------------
py-sphinx@1.3.1
vpn-128-154-224-57:docs me$ spack load py-sphinx
vpn-128-154-224-57:docs me$ make
sphinx-apidoc -T -o . ../../spack/spack
Traceback (most recent call last):
  File "/Users/me/spack/opt/spack/darwin-elcapitan-x86_64/clang-7.3.0-apple/py-sphinx-1.3.1-t4j2snyy2fwja5rknsbbp6mkf5norpbs/bin/sphinx-apidoc", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/pkg_resources.py", line 2793, in <module>
    working_set.require(__requires__)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/pkg_resources.py", line 673, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/pkg_resources.py", line 576, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: Sphinx==1.3.1
make: *** [apidoc] Error 1
```

**Comment by adamjstewart** [comment]
### Flake8

Flake8 isn't used for .rst files, it's used for PEP8 compliance of Python files. You really shouldn't need context for flake8 warnings. It usually just tells you that a specific line is too long or not indented properly. You can just add `$SPACK_ROOT/share/spack/qa` to your `PATH` and run `run-flake8-tests` and it will tell you whether you are good or not. Doesn't matter what directory you are in.

### Unit tests

To run all unit tests, simply run `spack test`. You can run a specific unit test with `spack test architecture`. Running `spack test --list` will give you a list of all available unit tests.

The `run-unit-tests` script basically runs a couple integration tests and then runs `spack test

**Comment by citibeth** [comment]
> Flake8 isn't used for .rst files, it's used for PEP8 compliance of Python files.

Sorry, I knew that... fat fingers late at night.

> You can just add $SPACK_ROOT/share/spack/qa to your PATH and run run-flake8-tests and it will tell you whether you are good or not. Doesn't matter what directory you are in.

This is good info for the manual.  It should probably document how to run on just one file, or all files.

> You really shouldn't need context for flake8 warnings. It usually just tells you that a specific line is too long or not indented properly.

The issue here is, some of the errors are "too many blank lines" or "missing blank lines."  Once you correct those errors, the line numbers on the other errors are wrong.  Unless you're good at keeping track of line number offsets, you need to re-run flake8.  Not a problem locally, but it does make it hard to correct flake8 errors reliably using the Travis logs.

Another problem is, often I introduce a new error when correcting the first.  That's one more reason to run `flake8` locally.

Finally... what is the preferred way to split string literals that are two long?  Or is there a way to tell flake8 to accept them?  Need Spack-specific guidance on this issue in your write-up...

> To run all unit tests, simply run spack test. You can run a specific unit test with
> spack test architecture. Running spack test --list will give you a list of all available
> unit tests.

That should work fine.

> Not sure what that error message means, weird. Are you using the Python installation that you built Sphinx with?

I'm using the system-installed Python on my Mac El Capitan, using Apple's default compilers.  This Spack setup is just to install software like `flake8`, `sphinx`, etc, not to build scientific stuff.  Here is `packages.yaml` (the tcl is in case I want to build environment-modules):
```
packages:
    python:
        paths:
            python@2.7.10: /usr
        version: [2.7.10]
        buildable: False
    tcl:
        paths:
            tcl@8.5: /usr
        version: [8.5]
        buildable: False
```
Here is `compilers.yaml`.  Looks like it detected my compiler twice, one instance of which I commented out:
```
compilers:
- compiler:
    modules: []
    operating_system: elcapitan
    paths:
      cc: /usr/bin/clang
      cxx: /usr/bin/clang++
      f77: null
      fc: null
    spec: clang@7.3.0-apple
#- compiler:
#    modules: []
#    operating_system: elcapitan
#    paths:
#      cc: /usr/bin/gcc
#      cxx: /usr/bin/g++
#      f77: null
#      fc: null
#    spec: gcc@4.2.1
```
And here's what I've installed:
```
$ spack find
==> 8 installed packages.
-- darwin-elcapitan-x86_64 / clang@7.3.0-apple ------------------
environment-modules@3.2.10  py-flake8@2.5.4  py-pil@1.1.7       py-setuptools@20.7.0
py-docutils@0.12            py-nose@1.3.7    py-pygments@2.0.2  py-sphinx@1.3.
```

**Comment by citibeth** [comment]
Another way to improve the QA stuff would be to add checks for common ``package.py`` problems.  Writing custom checks for these in Python is not too hard.  For example:
  * Lack of copyright notice
  * Left the FIXMEs in.

The guide should explain Spack's policy of copyright assignment (or lack thereof) for contributors.

**Comment by citibeth** [comment]
An update on the Sphinx stuff...

First problem, I was trying to use the system-installed Python on a Mac.  This is not recommended:
    https://hackercodex.com/guide/python-development-environment-on-mac-osx/

I then removed the system Python from my ``packages.yaml``, then do ``spack install sphinx``.  Now I ran into the issues that ``py-sphinx`` has a lot of stealth dependencies that are handled by PIP, not Spack.    Results from that are here: [spack.build.out.txt](https://github.com/LLNL/spack/files/453997/spack.build.out.txt)

The immediate problem is, PyPI seems to have certificate problems.  Who knows why.  I'm on NASA time, which seems to be 10 seconds ahead of offical US time.

http://stackoverflow.com/questions/16370583/pip-issue-installing-almost-any-library

The bigger problem is we have two auto-install systems both trying to fetch and install dependencies.  Maybe we "should" wrap each PyPI package as a Spack package and declare dependencies correctly.  Then our Spack packages would install OK.

Better in the long run would be to have a system that can auto-wrap this stuff on the fly.  The idea is: read a PyPI ``setup.py`` file, instantiate a Spack ``Package`` on the fly.  Then we could leverage off of PyPI instead of rewriting it all ourselves.

This issue needs more thought, IMHO.  In the meantime, MacPort's ``port install py-sphinx`` did the trick with the least amount of effort.  Apparently, they've manually wrapped all the PhPI packages.

**Comment by alalazo** [comment]
> Better in the long run would be to have a system that can auto-wrap this stuff on the fly. The idea is: read a PyPI setup.py file, instantiate a Spack Package on the fly. Then we could leverage off of PyPI instead of rewriting it all ourselves.

See #1249

**Comment by adamjstewart** [comment]
> Now I ran into the issues that py-sphinx has a lot of stealth dependencies that are handled by PIP, not Spack. 

@citibeth Are you using the latest Spack develop? `py-sphinx` now depends_on `py-sphinx-rtd-theme` as of #1563. If it's still trying to use pip to find that dependency, then we have other problems.

**Comment by citibeth** [comment]
I was able to use Sphinx to fix a lot of doc errors locally.  I was not able to get it fully build, due to linkages between the Spinx docs and running Spack.  This is not "plug-and-play."  Apparently, `hg` was needed to build, but I did manage to get that installed.

I pointed ReadTheDocs at my repo, and it was not able to build the docs successfully either.  It tried to run Spack, and didn't know how to.

https://readthedocs.org/projects/spack-citibeth-fork/builds/4390993/

This is a good argument to NOT use "auto-run" features from Sphinx, unless there's a good reason to do so.  The fewer dependencies the docs have, the better.  Also, instructions on hooking up Sphinx + Spack + hg + whatever are needed so users can re-create the Sphinx build on their own machines.

**Comment by adamjstewart** [comment]
Yeah, if you run `run-doc-tests`, it checks for any missing dependencies first.

**Comment by citibeth** [comment]
On Sat, Sep 10, 2016 at 2:30 PM, Adam J. Stewart <notifications@github.com>
wrote:

> Yeah, if you run run-doc-tests, it checks for any missing dependencies
> first.
>

It misses some dependencies, see below.

```
$ share/spack/qa/run-doc-tests
Dependencies found.
sphinx-apidoc -T -o . ../../spack/spack
Creating file ./spack.rst.
Creating file ./spack.cmd.rst.
Creating file ./spack.cmd.common.rst.
Creating file ./spack.compilers.rst.
Creating file ./spack.hooks.rst.
Creating file ./spack.operating_systems.rst.
Creating file ./spack.platforms.rst.
Creating file ./spack.schema.rst.
Creating file ./spack.test.rst.
Creating file ./spack.test.cmd.rst.
Creating file ./spack.util.rst.
sphinx-build -b html -d _build/doctrees  -W . _build/html
Running Sphinx v1.4.5
making output directory...
loading pickled environment... not yet created
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 22 source files that are out of date
updating environment: 22 added, 0 changed, 0 removed
reading sources... [100%] workflows

looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [  4%] basic_usage

Warning, treated as error:
WARNING: dot command 'dot' cannot be run (needed for graphviz output),
check the graphviz_dot setting
```

**Comment by adamjstewart** [comment]
See #1691.

**Comment by citibeth** [comment]
A few more issues with `run-doc-tests`:

1. It doesn't work when run from an arbitrary directory.
2. It doesn't leave the HTML files behind, without hacking the script.  There's no "out-of-the-box" way to build the manual
3. There's no easy way to put the manual up on readthedocs.

**Comment by adamjstewart** [comment]
1. Yes it does? I just tried it and it works even if you aren't in the Spack repo.
2. I did that on purpose, but maybe I should change that. If you want to build the manual out of the box, just run `make` in the docs directory.
3. I agree, it would be nice to see the changes in HTML instead of in .rst.

**Comment by adamjstewart** [comment]
I just noticed that if you run `make` multiple times without running `make clean` in the middle, it generates some warning messages. I wonder if there's a way to fix that. I should probably run `make clean` at the beginning of `run-doc-tests` just to be safe.

**Comment by citibeth** [comment]
When warnings are treated as errors, Sphinx seems to stop after the first one.  This is inefficient when trying to clean up problems.

**Comment by adamjstewart** [comment]
Yes, I set `SPHINXOPTS=-W` to treat all warnings as fatal errors. If I didn't do this, the exit code of make would be 0. To be honest, `run-doc-tests` is really for Travis CI. If you want to see a list of all warning messages, run `make` yourself.

**Comment by citibeth** [comment]
>
>    1. I did that on purpose, but maybe I should change that. If you want
>    to build the manual out of the box, just run make in the docs
>    directory.
>
>
That doesn't work, it can't find the Python code it needs.  The env vars
set in run-docs-test are needed, I believe.

**Comment by adamjstewart** [comment]
> That doesn't work, it can't find the Python code it needs.  The env vars
set in run-docs-test are needed, I believe.

How recent is your fork? I think @tgamblin fixed most of those errors by setting those env vars in conf.py:
https://github.com/LLNL/spack/commits/develop/lib/spack/docs/conf.py

If rebasing doesn't fix it, can you post the error messages?
